llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp

```diff
@@ -528,14 +528,6 @@ class SCEVAddRecForUniformityRewriter
     return SCEVRewriteVisitor<SCEVAddRecForUniformityRewriter>::visit(S);
   }
 
-  const SCEV *visitUnknown(const SCEVUnknown *S) {
```

Comment:
Hmm, I don't think removing all checks for SCEVUnknown is safe.

The rewriter is used to rewrite a SCEV expression for each lane, and checks if the result is the same SCEV for all lanes.

I think with this change, SCEVUnknown would always be treated as uniform, but they could be any load instruction for example, which may or may not read a different value from memory on every iteration? (needs a test case I think).

We might be able to refine the checks to handle additional cases, e.g. read-only loops?

Reply:
Thanks for the example: I _think_ the transformation performed is correct? I seem to be having problems with Alive2: https://alive2.llvm.org/ce/z/CTS75q

Reply:
Fixed the proof: it is incorrect. https://alive2.llvm.org/ce/z/EgYxUg

Reply:
Thanks a lot for the example, by the way: it is somewhat unfortunate that we're blocked on the SCEVUnknown of the load, behind which an SCEVAddRec is hiding. I wonder if we can still get the benefits of the patch by looking inside the SCEVUnknown, and asserting that it is uncomputable if it is an Instruction?